codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttEncoder.java

```diff
@@ -196,138 +230,270 @@ private static int getConnVariableHeaderFlag(MqttConnectVariableHeader variableH
+                String topicName = topic.topicName();
+                byte[] topicNameBytes = encodeStringUtf8(topicName);
+                payloadBufferSize += 2 + topicNameBytes.length;
+                payloadBufferSize += 1;
+            }
 
-        // Variable Header
-        int messageId = variableHeader.messageId();
-        buf.writeShort(messageId);
+            int variablePartSize = variableHeaderBufferSize + payloadBufferSize;
+            int fixedHeaderBufferSize = 1 + getVariableLengthInt(variablePartSize);
+
+            ByteBuf buf = ctx.alloc().buffer(fixedHeaderBufferSize + variablePartSize);
+            buf.writeByte(getFixedHeaderByte1(mqttFixedHeader));
+            writeVariableLengthInt(buf, variablePartSize);
+
+            // Variable Header
+            int messageId = variableHeader.messageId();
+            buf.writeShort(messageId);
+            buf.writeBytes(propertiesBuf);
+
+            // Payload
+            for (MqttTopicSubscription topic : payload.topicSubscriptions()) {
+                writeUTF8String(buf, topic.topicName());
+                final MqttSubscriptionOption option = topic.option();
+
+                int optionEncoded = option.retainHandling().value() << 4;
+                if (option.isRetainAsPublished()) {
+                    optionEncoded |= 0x08;
+                }
+                if (option.isNoLocal()) {
+                    optionEncoded |= 0x04;
+                }
+                optionEncoded |= option.qos().value();
+
+                buf.writeByte(optionEncoded);
+            }
 
-        // Payload
-        for (MqttTopicSubscription topic : payload.topicSubscriptions()) {
-            String topicName = topic.topicName();
-            byte[] topicNameBytes = encodeStringUtf8(topicName);
-            buf.writeShort(topicNameBytes.length);
-            buf.writeBytes(topicNameBytes, 0, topicNameBytes.length);
-            buf.writeByte(topic.qualityOfService().value());
+            return buf;
+        } finally {
+            propertiesBuf.release();
         }
-
-        return buf;
     }
 
     private static ByteBuf encodeUnsubscribeMessage(
-            ByteBufAllocator byteBufAllocator,
+            ChannelHandlerContext ctx,
             MqttUnsubscribeMessage message) {
-        int variableHeaderBufferSize = 2;
-        int payloadBufferSize = 0;
-
-        MqttFixedHeader mqttFixedHeader = message.fixedHeader();
-        MqttMessageIdVariableHeader variableHeader = message.variableHeader();
-        MqttUnsubscribePayload payload = message.payload();
+        MqttVersion mqttVersion = MqttCodecUtil.getMqttVersion(ctx);
+        ByteBuf propertiesBuf = encodePropertiesIfNeeded(mqttVersion,
+                ctx.alloc(),
+                message.idAndPropertiesVariableHeader().properties());
+
+        try {
+            final int variableHeaderBufferSize = 2 + propertiesBuf.readableBytes();
```

Comment:
There's a lot of integers in this file, 1, 2, 4, 0x04, 0x08, 0x01, etc. Would be nice if they were named constants.

Reply:
Eventually, this may be a nice improvement. I think, better do it as a separate PR - refactoring can be done infinitely long :)